packages/dtslint/src/index.ts

```diff
@@ -142,10 +157,77 @@ async function runTests(
+  let hasNpmVersionMismatch = false;
+
+  if (!skipNpmChecks && !expectOnly) {
+    const attw = await import("@arethetypeswrong/core");
+    const typesPackageVersion = `${packageJson.libraryMajorVersion}.${packageJson.libraryMinorVersion}`;
+    const pkg = await tryPromise(
+      attw.resolveImplementationPackageForTypesPackage(packageJson.name, `${typesPackageVersion}.9999`),
+    );
+    if (pkg) {
+      const { packageName, packageVersion, tarballUrl } = pkg;
+      if (packageJson.nonNpm === true) {
+        (errors ??= []).push(
+          `Package ${packageJson.name} is marked as non-npm, but ${packageName} exists on npm. ` +
+            `If these types are being added to DefinitelyTyped for the first time, please choose ` +
+            `a different name that does not conflict with an existing npm package.`,
+        );
+      } else if (!packageJson.nonNpm) {
+        if (!satisfies(packageVersion, typesPackageVersion)) {
+          hasNpmVersionMismatch = true;
+          const isError = !npmVersionExemptions.has(packageDirectoryNameWithVersion);
+          const container = isError ? (errors ??= []) : (warnings ??= []);
+          container.push(
+            (isError
+              ? ""
+              : `Ignoring npm version error because ${packageDirectoryNameWithVersion} was failing when the check was added. ` +
+                `If you are making changes to this package, please fix this error:\n> `) +
+              `Cannot find a version of ${packageName} on npm that matches the types version ${typesPackageVersion}. ` +
+              `The closest match found was ${packageName}@${packageVersion}. ` +
+              `If these types are for the existing npm package ${packageName}, change the ${packageDirectoryNameWithVersion}/package.json ` +
+              `major and minor version to match an existing version of the npm package. If these types are unrelated to ` +
+              `the npm package ${packageName}, add \`"nonNpm": true\` to the package.json and choose a different name ` +
+              `that does not conflict with an existing npm package.`,
+          );
+        } else {
+          const tarballPath = joinPaths(tmpDir, `${packageName.replace(/\//g, "-")}-${packageVersion}.tgz`);
+          const unpackedPath = joinPaths(tmpDir, `${packageName}-${packageVersion}`);
+          await pipeline((await fetch(tarballUrl)).body!, fs.createWriteStream(tarballPath));
+          await fs.promises.mkdir(unpackedPath, { recursive: true });
+          await pipeline(fs.createReadStream(tarballPath), createGunzip(), tar.extract({ cwd: unpackedPath }));
+          const containerDir = (await fs.promises.readdir(unpackedPath))[0];
+          implementationPackage = {
+            packageName,
+            packageVersion,
+            tarballPath,
+            unpackedPath: joinPaths(unpackedPath, containerDir),
+          };
+        }
+      }
+    } else {
+      if (packageJson.nonNpm === "conflict") {
```

Comment:
it should be equivalent to flatten this into a if/else if/else if sequence.

```suggestion
    } else if (packageJson.nonNpm === "conflict") {
```